account_payment_order/models/account_move.py

```diff
@@ -0,0 +1,241 @@
+# © 2014 Serv. Tecnol. Avanzados - Pedro M. Baeza
+# © 2016 Akretion (Alexis de Lattre <alexis.delattre@akretion.com>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    payment_order_id = fields.Many2one(
+        comodel_name="account.payment.order",
+        string="Payment Order",
+        copy=False,
+        readonly=True,
+        check_company=True,
+    )
+    payment_order_ok = fields.Boolean(compute="_compute_payment_order_ok")
+    # we restore this field from <=v11 for now for preserving behavior
+    # TODO: Check if we can remove it and base everything in something at
+    # payment mode or company level
+    reference_type = fields.Selection(
+        selection=[("none", "Free Reference"), ("structured", "Structured Reference")],
+        readonly=True,
+        default="none",
+    )
+    payment_line_count = fields.Integer(compute="_compute_payment_line_count")
+
+    @api.depends("payment_mode_id", "line_ids", "line_ids.payment_mode_id")
+    def _compute_payment_order_ok(self):
+        for move in self:
+            payment_mode = move.line_ids.filtered(lambda x: not x.reconciled).mapped(
+                "payment_mode_id"
+            )[:1]
+            if not payment_mode:
+                payment_mode = move.payment_mode_id
+            move.payment_order_ok = payment_mode.payment_order_ok
+
+    def _compute_payment_line_count(self):
+        for move in self:
+            move.payment_line_count = len(
+                self.env["account.payment.line"]._search(
+                    [("move_line_id", "in", self.line_ids.ids)]
+                )
+            )
+
+    def _get_payment_order_communication_direct(self):
+        """Retrieve the communication string for this direct item."""
+        communication = self.payment_reference or self.ref or self.name
+        if self.is_invoice():
+            if self.is_purchase_document():
+                communication = self.ref or self.payment_reference
+            else:
+                communication = self.payment_reference or self.name
+        return communication or ""
+
+    def _get_payment_order_communication_full(self):
+        """Retrieve the full communication string for the payment order.
+        Reversal moves and partial payments references added.
+        Avoid having everything in the same method to avoid infinite recursion
+        with partial payments.
+        """
+        communication = self._get_payment_order_communication_direct()
+        references = []
+        # Build a recordset to gather moves from which references have already
+        # taken in order to avoid duplicates
+        reference_moves = self.env["account.move"].browse()
+        # If we have credit note(s) - reversal_move_id is a one2many
+        if self.reversal_move_id:
```

Comment:
This field no longer exists; it was renamed to `reversed_entry_id`.

```suggestion
        if self.reversed_entry_id:
```

![image](https://github.com/user-attachments/assets/111050ff-5b36-4145-bcac-8b63ccb9a99a)